### PR TITLE
Runtime estimator

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Estimates remaining runtime. Uses a simple timer interpolation + correction for eval. Graph shows a "normal" run and a corrected run with eval.

image

Note that this doesn't always work super smoothly, e.g. resnet estimates are not great because first epoch is a lot slower due to dataloader issues.

image

We will probably hit more issues as well later on, so this will require corrections.

Import Questions:
- is there a different place to log this? `wall_clock/` is intuitive, but everything else is in seconds so we need to report in seconds (whereas we would prefer to probably report hours?)

# What issue(s) does this change relate to?

[CO-1664](https://mosaicml.atlassian.net/browse/CO-1664)

[CO-1664]: https://mosaicml.atlassian.net/browse/CO-1664?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mvpatel2000** [comment]
Note: The weird spikes in the wandb graph are because wandb associates the last logged time for a given step, which makes graph weird when x-axis is time. I spent a while investigating -- happy to discuss offline since it's a little too complicated to writeup. As a side effect, @eracah and I discovered where timestamp is advanced for batches vs. epochs compared to when metrics are calculated is inconsistent -- along for the ride I reorder the epoch one to be consistent.

**Review by Glabred — CHANGES_REQUESTED** [review]
say sorry to me

**Review by Glabred — COMMENTED** [review]
jk

File: composer/callbacks/runtime_estimator.py
Comment (dakinggg) [reply composer/callbacks/runtime_estimator.py:103]: didn't comment on all of them, but there are a bunch of prints that should probably be logs?

File: composer/callbacks/runtime_estimator.py
Comment (dakinggg) [inline composer/callbacks/runtime_estimator.py:139]:
```suggestion
                # Discard first eval_wct if possible as it is often slower due to dataset downloading
```

File: composer/callbacks/runtime_estimator.py
Comment (dakinggg) [inline composer/callbacks/runtime_estimator.py:172]: is this possible to hit? batch_start should set `self._enabled` to `False` if `max_duration` is `None`, so I think this would never be hit?

File: composer/callbacks/runtime_estimator.py
Comment (dakinggg) [inline composer/callbacks/runtime_estimator.py:150]: Is it possible for `eval_rate` to be zero? I would think not?

**Review by dakinggg — APPROVED** [review]
LGTM with some comments

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:2035]: nvm, read your comment after reviewing.

**Comment by dakinggg** [comment]
Also, I'm willing to merge this with just manual tests, but please do add some unit tests in a follow on pr

File: tests/common/models.py
Comment (bcui19) [inline tests/common/models.py:7]: Do you want to move these files to another PR -- they seem related to another task?

**Review by bcui19 — APPROVED** [review]
LGTM, after addressing comments. +1 to adding tests in a separate PR

File: tests/common/models.py
Comment (mvpatel2000) [reply tests/common/models.py:7]: This is because the test is failing and blocking this PR :(

**Comment by mvpatel2000** [comment]
Added a test

**Review by hanlint — APPROVED** [review]

**Pull request merged**
